modules/examples/sources/threaded-random-generator/threaded-random-generator.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "threaded-random-generator.h"
+#include "logthrsource/logthrsourcedrv.h"
+#include "atomic.h"
+#include "messages.h"
+#include "str-format.h"
+
+#include <strings.h>
+#include <sys/random.h>
+#include <unistd.h>
+
+struct ThreadedRandomGeneratorSourceDriver
+{
+  LogThreadedSourceDriver super;
+  GAtomicCounter exit_requested;
+
+  guint freq;
+  guint bytes;
+  gboolean urandom;
+};
+
+static gboolean
+_generate_random_bytes(guint8 *random, guint size, gboolean urandom)
+{
+  guint length = 0;
+
+  while (length < size)
+    {
+      gssize rc = getrandom(random + length, size - length, urandom ? 0 : GRND_RANDOM);
+
+      if (rc < 0)
+        {
+          msg_error("Could not generate random bytes", evt_tag_error("error"));
+          return FALSE;
+        }
+
+      length += rc;
+    }
+
+  return TRUE;
+}
+
+
+/* runs in a dedicated thread */
+static void
+_run(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  guint8 *random_bytes = g_malloc(self->bytes);
+
+  const gsize random_hex_str_size = self->bytes * 2 + 1;
+  gchar *random_hex_str = g_malloc(random_hex_str_size);
+
+  while (!g_atomic_counter_get(&self->exit_requested))
+    {
```

Comment:
Like I mentioned in the LogThreadedSourceDriver stuff, I think it would be better if these kind of exit strategies could be implemented by the source driver framework alone.

e.g. log_threaded_source_driver_should_exit() or similar call could be used to wrap the g_atomic logic we have now.

Reply:
I think it is not possible, if we want to support the integration of external event loops.

For example, if one wants to start a `libuv` event loop in C, or a Boost.Asio loop in C++, maybe a Flask/whatever server from a Python source. Their API require `request_exit` as a "notification". It would be either hard or ugly to add a `log_threaded_source_driver_should_exit()` polling call to this use case.

Reply:
If possible, try to modify the test and use `libuv` ;)